app/gui2/src/components/visualizations/ScatterplotVisualization.vue

```diff
@@ -563,6 +565,82 @@ function getPlotData(data: Data) {
   return data.data
 }
 
+const filterPattern = computed(() => Pattern.parse('__ (..Between __ __)'))
+const makeFilterPattern = (
+  module: Ast.MutableModule,
+  columnName: string,
+  min: number,
+  max: number,
+) => {
+  return filterPattern.value.instantiateCopied([
+    Ast.TextLiteral.new(columnName),
+    Ast.tryNumberToEnso(min, module)!,
+    Ast.tryNumberToEnso(max, module)!,
+  ])
+}
+function getAstPatternFilterAndSort(
+  series: string[],
+  xColName: string,
+  minX: number,
+  maxX: number,
+  minY: number,
+  maxY: number,
+) {
+  return Pattern.new((ast) => {
+    let pattern = Ast.OprApp.new(
+      ast.module,
+      Ast.App.positional(
+        Ast.PropertyAccess.new(ast.module, ast, Ast.identifier('filter')!),
+        makeFilterPattern(ast.module, xColName, minX, maxX),
+      ),
+      '.',
+      Ast.App.positional(
+        Ast.Ident.new(ast.module, Ast.identifier('filter')!),
+        makeFilterPattern(ast.module, series[1]!, minY, maxY),
+      ),
+    )
+    for (const s in series) {
+      pattern = Ast.OprApp.new(
+        ast.module,
+        pattern,
+        '.',
+        Ast.App.positional(
+          Ast.Ident.new(ast.module, Ast.identifier('filter')!),
+          makeFilterPattern(ast.module, series[s]!, minY, maxY),
+        ),
+      )
+    }
+    return pattern
+  })
+}
+const createNewFilterNode = () => {
+  if (!data.value.isTimeSeries) {
+    const seriesLabels = Object.keys(data.value.axis)
+      .filter((s) => s != 'x')
+      .map((s) => {
+        return data.value.axis[s as keyof AxesConfiguration].label
+      })
+    const xColName = data.value.axis.x.label
+    const xItems = data.value.data.map((d) => d.x)
```

Comment:
To avoid the casts below
```suggestion
    const xItems = data.value.data.map((d) => d.x as number)
```